testing/logging/test_reporting.py

```diff
@@ -1165,3 +1165,18 @@ def test_log_file_cli_subdirectories_are_successfully_created(
     result = pytester.runpytest("--log-file=foo/bar/logf.log")
     assert "logf.log" in os.listdir(expected)
     assert result.ret == ExitCode.OK
+
+
+def test_locationline_returns_best_relative_location(pytester: Pytester) -> None:
+    item = pytester.getitem("def test_func(): pass")
+    item.config.option.verbose = 2
+
+    tr = TerminalReporter(item.config)
+    path_to_test = os.path.join("unique", "path")
+    fspath = os.path.join(pytester.path, path_to_test)
+    nodeid = "nodeid::test"
+
+    result = tr._locationline(nodeid, fspath, 1, "domain")
```

Comment:
So I couldn't come up with a test that will trigger the problematic code path. I think it can be done on a Windows machine where the test and the rootdir or on different drives, but I don't use Windows.

So my suggestion is to "force" this by having `bestrelpath` assert the types of its arguments:

```diff
index f641e6491..22b7ebd75 100644
--- a/src/_pytest/pathlib.py
+++ b/src/_pytest/pathlib.py
@@ -685,6 +685,9 @@ def bestrelpath(directory: Path, dest: Path) -> str:
 
     If no such path can be determined, returns dest.
     """
+    assert isinstance(directory, Path)
+    assert isinstance(dest, Path)
+
     if dest == directory:
         return os.curdir
     # Find the longest common directory.
```

This will cause some tests to fail anyway (so this test shouldn't be needed), even though they have a common base path. It slows down things a bit, but I think it's good to avoid hidden bugs like this.

Reply:
Hey, cheers! The asserts have been added, but to confirm, is this a test that should be in the main code? Just wondered as it will assert the type but not sure how that would help the end user? However, it is better than the original error that was raised when this function was passed a string.

Reply:
> Just wondered as it will assert the type but not sure how that would help the end user?

It will make sure our tests catch any misuse of the function. The problem is that because the `if not base:` condition rarely fails and depends on user input, and only the part after it requires `dest` to be `Path` (rather than `str`), it's easy to get by with passing `str` without noticing. The assert makes it visible.

> and I guess that won't happen at all now the type annotation exists.

Yep, as long as we don't miss any :)